DragaliaAPI/Controllers/Dragalia/FortController.cs

```diff
@@ -43,13 +70,695 @@ await this.fortRepository.GetBuilds(this.DeviceAccountId).ToListAsync()
+        );
+
+        // Move building to requested coordinate
+        build.PositionX = request.after_position_x;
+        build.PositionZ = request.after_position_z;
+        this.fortRepository.UpdateBuild(build);
+
+        UpdateDataList updateDataList = this.updateDataService.GetUpdateDataList(
+            this.DeviceAccountId
+        );
+
+        await this.fortRepository.SaveChangesAsync();
+
+        FortMoveData data =
+            new()
+            {
+                result = 1,
+                build_id = request.build_id,
+                fort_bonus_list = bonusList,
+                production_rp = StubData.ProductionRp,
+                production_df = StubData.ProductionDf,
+                production_st = StubData.ProductionSt,
+                update_data_list = updateDataList
+            };
+        return this.Ok(data);
+    }
+
+    // What exactly does it do
+    [HttpPost("set_new_fort_plant")]
+    public async Task<DragaliaResult> SetNewFortPlant(FortSetNewFortPlantRequest request)
+    {
+        await this.fortRepository.GetFortPlantIdList(request.fort_plant_id_list);
+        
+        UpdateDataList updateDataList = this.updateDataService.GetUpdateDataList(
+            this.DeviceAccountId
+        );
+        
+        FortSetNewFortPlantData data =
+            new()
+            {
+                result = 1,
+                update_data_list = updateDataList
+            };
+        return this.Ok(data);
+    }
+
+    private async Task IncrementCarpenterUsage(FortDetail fortDetail)
+    {
+        fortDetail.working_carpenter_num++;
+        await this.fortRepository.UpdateFortWorkingCarpenter(
+            this.DeviceAccountId, 
+            fortDetail.working_carpenter_num
+        );
+    }
+
+    private async Task DecrementCarpenterUsage(FortDetail fortDetail)
+    {
+        fortDetail.working_carpenter_num--;
+        if (fortDetail.working_carpenter_num < 0)
+        {
+            fortDetail.working_carpenter_num = 0;
+        }
+
+        await this.fortRepository.UpdateFortWorkingCarpenter(
+            this.DeviceAccountId, 
+            fortDetail.working_carpenter_num
+        );
+    }
+
+    private async Task ConsumePlayerMaterials(IQueryable<DbPlayerMaterial> userMaterials, DbPlayerUserData userData, FortPlantDetail plantDetail)
+    {
+        userData.Coin -= plantDetail.Cost;
+
+        // Is there maybe a more efficient way to do this? Seems ugly, but works
+        await ConsumeMaterial(userMaterials, plantDetail.MaterialsId1, plantDetail.MaterialsNum1);
+        await ConsumeMaterial(userMaterials, plantDetail.MaterialsId2, plantDetail.MaterialsNum2);
+        await ConsumeMaterial(userMaterials, plantDetail.MaterialsId3, plantDetail.MaterialsNum3);
+        await ConsumeMaterial(userMaterials, plantDetail.MaterialsId4, plantDetail.MaterialsNum4);
+        await ConsumeMaterial(userMaterials, plantDetail.MaterialsId5, plantDetail.MaterialsNum5);
```

Comment:
there is indeed, have a look at https://github.com/SapiensAnatis/DragaliaAPI/blob/develop/DragaliaAPI.Database/Repositories/InventoryRepository.cs#L158

In the weapons case I have built a map of materials in the MasterAsset model: https://github.com/SapiensAnatis/DragaliaAPI/blob/develop/DragaliaAPI.Shared/MasterAsset/Models/WeaponBodyBuildupGroup.cs#L29 and just deduct that

Reply:
Very nifty, managed to adapt something very similar to this case, thanks!